provider/datagen/src/driver.rs

```diff
@@ -950,88 +931,111 @@ fn select_locales_for_key(
+                    debug_assert_eq!(*family, LocaleFamily::full());
+                    include_full = true;
+                    continue;
+                };
+                let value = locales_map.entry(langid.clone()).or_default();
+                value.is_selected = true;
+                if *langid != LanguageIdentifier::UND {
+                    value.family = Some(family);
+                }
             }
-            if locale.is_langid_und() && include_und {
-                return true;
+            if include_full && families.len() == 1 {
+                // Special case: return now so we don't need the fallbacker
+                let selected_locales = locales_map
+                    .into_iter()
+                    .flat_map(|(_, value)| value.data_locales)
+                    .collect();
+                return Ok(selected_locales);
             }
-            if locale.language().is_empty()
-                && matches!(
-                    key.fallback_config().priority,
-                    icu_provider::FallbackPriority::Region
-                )
-            {
-                return true;
+        }
+        LocalesWithOrWithoutFallback::WithoutFallback { langids: locales } => {
+            for langid in locales {
+                let value = locales_map.entry(langid.clone()).or_default();
+                value.is_selected = true;
             }
-            // Special case: skeletons *require* the -u-ca keyword, so don't export locales that don't have it
-            // This would get caught later on, but it makes datagen faster and quieter to catch it here
-            if key.path().get() == "datetime/skeletons@1" && !locale.has_unicode_ext() {
-                return false;
+        }
+    }
+
+    // Fill in missing extensions and aux keys from parent locales,
+    // and calculate which langids are ancestors and descendants.
+    for current_langid in locales_map.keys().cloned().collect::<Vec<_>>() {
+        let current_value = locales_map.get_mut(&current_langid).unwrap();
+        if include_full && !current_value.is_selected {
+            log::trace!("Including {current_langid}: the full locale family is present");
+            current_value.is_selected = true;
+        }
+        if current_langid.language.is_empty() {
+            if current_langid != LanguageIdentifier::UND {
+                log::trace!("Including {current_langid}: variants of und are always included");
+                current_value.is_selected = true;
             }
-            let mut iter = fallbacker_with_config.fallback_for(locale);
-            while !iter.get().is_und() {
-                if explicit.contains(iter.get()) {
-                    return true;
+            // Nothing left to be done here
+            continue;
+        }
+        let include_ancestors = current_value
+            .family
+            .map(|family| family.include_ancestors)
+            .unwrap_or(false);
```

Comment:
Is is really worth, performance wise, to have family be an optional borrow, instead of just creating the families for all LanguageIdentifiers? This `false` is the same as `single`'s, but here it's just a `false` literal.

Same below for `include_descendants`.

Reply:
The presence of the family is how I indicate that the langid was explicit. If it is `None`, the langid is supported but not explicit.